### PR TITLE
Added recipe for writing test with code depending on a window object

### PR DESCRIPTION
**Pull request by bachstatter**
I was debating if it should be called react-testing or browser-testing and decided to call it browser-testing since if solves a problem for all libraries that expect a global window object.

I'm happy to add changes based on feedback

### CODE REVIEW

File: docs/recipes/browser-testing.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# Setting up AVA for browser testing.
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Extra "." here

File: docs/recipes/browser-testing.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+Ava doesn't come with support for the browser.
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: "Ava" -> "AVA"

Comment (vdemedes) [inline docs/recipes/browser-testing.md]: "AVA does not support running tests in browsers yet." - maybe

File: docs/recipes/browser-testing.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+Some libraries require that a window object is defined.
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: "window" -> `window`

Comment (vdemedes) [inline docs/recipes/browser-testing.md]: "Some libraries require `window` object to be defined."

File: docs/recipes/browser-testing.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+An example of this is React.js, at least if you want to use ReactDOM.render and simulate events with ReactTestUtils.
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: "React.js" -> "React"

File: docs/recipes/browser-testing.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+This recipe will work for any library that needs a mocked browser environment.
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: "This recipe works"

File: docs/recipes/browser-testing.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+## First step
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Instead of first/second/Nth step, better provide a meaningful title that indicates what this step is about. For example, "Install jsdom"

File: docs/recipes/browser-testing.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+Install [jsdom](https://github.com/tmpvar/jsdom)
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Missing dot.

File: docs/recipes/browser-testing.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+> A JavaScript implementation of the WHATWG DOM and HTML standards, for use with node.js
+
+```
+$ npm install jsdom --save-dev
+```
+
+## Second step
+Create a helper file. Ava will ignore to process all files in a folder called helpers.
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Ava -> AVA

File: docs/recipes/browser-testing.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+## First step
+Install [jsdom](https://github.com/tmpvar/jsdom)
+
+> A JavaScript implementation of the WHATWG DOM and HTML standards, for use with node.js
+
+```
+$ npm install jsdom --save-dev
+```
+
+## Second step
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Set up jsdom

File: docs/recipes/browser-testing.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+Create a helper file. Ava will ignore to process all files in a folder called helpers.
+
+`test/helpers/setup-browser-env.js`:
+```js
+global.document = require('jsdom').jsdom('<body></body>')
+global.window = document.defaultView
+global.navigator = window.navigator
+
+// Transpile the files we import/require as well, optional
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: I think this part is out of scope of this recipe.

File: docs/recipes/browser-testing.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+require('babel-core/register')
+// Uncomment if you want have hot module reloading and want to avoid it
+// process.env.BABEL_ENV = 'test'
+```
+
+## Third step
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Configure all tests to set up jsdom, or something like that

File: docs/recipes/browser-testing.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+{
+  "ava": {
+    "require": [
+      "./test/helpers/setup-browser-env.js"
+    ]
+  }
+}
+```
+
+## Final step
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Enjoy!

File: docs/recipes/browser-testing.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+Write your tests and enjoy a mocked window object.
+(This example requires some more npm install)
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: Missing dot

File: docs/recipes/browser-testing.md [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+Ava doesn't come with support for the browser.
+Some libraries require that a window object is defined.
+An example of this is React.js, at least if you want to use ReactDOM.render and simulate events with ReactTestUtils.
+
+This recipe will work for any library that needs a mocked browser environment.
+
+## First step
+Install [jsdom](https://github.com/tmpvar/jsdom)
+
+> A JavaScript implementation of the WHATWG DOM and HTML standards, for use with node.js
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: node.js -> Node.js

File: docs/recipes/browser-testing.md (hunk 8 above)
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: folder called "helpers"

File: docs/recipes/browser-testing.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+// Transpile the files we import/require as well, optional
+require('babel-core/register')
+// Uncomment if you want have hot module reloading and want to avoid it
+// process.env.BABEL_ENV = 'test'
+```
+
+## Third step
+
+Make sure AVA requires our setup file before running any tests.
```
Comment (vdemedes) [inline docs/recipes/browser-testing.md]: requires a setup file (no "our")

**Comment by vdemedes** [comment]
We had some users ask for this one previously. I like it, thank you for your work @bachstatter! If you could address fixes/suggestions, it'd be good to merge ;)

File: docs/recipes/browser-testing.md (hunk 8 above)
Comment (novemberborn) [inline docs/recipes/browser-testing.md]: Perhaps:

> Create a helper file and place it in the `test/helpers` folder. This ensures AVA does not treat it as a test.

File: docs/recipes/browser-testing.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  const input = render(
+    React.createElement(CustomInput, {onUserBlur),
+    div,
+  )
+
+  Simulate.blur(input)
+
+  t.true(onUserBlur.calledOnce)
+})
+```
```
Comment (sindresorhus) [inline docs/recipes/browser-testing.md]: semi-colons please

File: docs/recipes/browser-testing.md [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+```js
+global.document = require('jsdom').jsdom('<body></body>')
+global.window = document.defaultView
+global.navigator = window.navigator
+
+// Transpile the files we import/require as well, optional
+require('babel-core/register')
+// Uncomment if you want have hot module reloading and want to avoid it
+// process.env.BABEL_ENV = 'test'
+```
```
Comment (sindresorhus) [inline docs/recipes/browser-testing.md]: semicolons

File: docs/recipes/browser-testing.md [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+This recipe will work for any library that needs a mocked browser environment.
+
+## First step
+Install [jsdom](https://github.com/tmpvar/jsdom)
+
+> A JavaScript implementation of the WHATWG DOM and HTML standards, for use with node.js
+
+```
+$ npm install jsdom --save-dev
```
Comment (sindresorhus) [inline docs/recipes/browser-testing.md]:
```
$ npm install --save-dev jsdom
```

**Comment by bachstatter** [comment]
Thanks for all the feedback.  English is my second language. Will update as soon as possible

File: docs/recipes/browser-testing.md (hunk 14 above)
Comment (bachstatter) [inline docs/recipes/browser-testing.md]: This is a quote from the jsdom repo description.

File: docs/recipes/browser-testing.md [hunk 19]
```diff
@@ -0,0 +1,69 @@
+# Setting up AVA for browser testing
+
+AVA does not support running tests in browsers yet.
+Some libraries require ```window``` object is defined.
```
Comment (jamestalmage) [inline docs/recipes/browser-testing.md]: This line could be better:

> Some libraries require browser specific globals (`window`, `document`, `navigator`, etc).

Also, You just need a single tick around `window` (not triple).

File: docs/recipes/browser-testing.md [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+`test/helpers/setup-browser-env.js`:
+```js
+global.document = require('jsdom').jsdom('<body></body>');
+global.window = document.defaultView;
+global.navigator = window.navigator;
+
+```
+
+## Configure all test to use jsdom
```
Comment (jamestalmage) [inline docs/recipes/browser-testing.md]: > Configure tests to use jsdom.

test => tests

"all" seems extraneous

File: docs/recipes/browser-testing.md [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+      "./test/helpers/setup-browser-env.js"
+    ]
+  }
+}
+```
+
+## Enjoy
+
+Write your tests and enjoy a mocked window object.
+(This example requires some more npm install.)
```
Comment (jamestalmage) [inline docs/recipes/browser-testing.md]: Drop this line. People can figure out they need to `npm install`

File: docs/recipes/browser-testing.md [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+```js
+global.document = require('jsdom').jsdom('<body></body>');
+global.window = document.defaultView;
+global.navigator = window.navigator;
+
+```
+
+## Configure all test to use jsdom
+
+Make sure AVA requires a setup file before running any tests.
```
Comment (jamestalmage) [inline docs/recipes/browser-testing.md]: Configure AVA to `require` the helper before every test file.

File: docs/recipes/browser-testing.md [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+Install [jsdom](https://github.com/tmpvar/jsdom).
+
+> A JavaScript implementation of the WHATWG DOM and HTML standards, for use with node.js
+
+```
+$ npm install --save-dev jsdom
+```
+
+## Setup jsdom
+Create a helper file and place it in the ```test/helpers``` folder. This ensures AVA does not treat it as a test.
```
Comment (jamestalmage) [inline docs/recipes/browser-testing.md]: triple tick to single around `test/helpers`

**Comment by jamestalmage** [comment]
LGTM. Just some minor nits.

File: docs/recipes/browser-testing.md [hunk 24]
```diff
@@ -0,0 +1,69 @@
+# Setting up AVA for browser testing
+
+AVA does not support running tests in browsers yet.
```
Comment (jamestalmage) [inline docs/recipes/browser-testing.md]: Link the word `yet` to the issue tracking browser support?

Comment (bachstatter) [inline docs/recipes/browser-testing.md]: :+1:

**Comment by sindresorhus** [comment]
Thank you @bachstatter. Good stuff :)
